File: internal/agentcore/subagent.go
package agentcore

import "time"

// SubAgentLink records that a Turn in a parent Session spawned a child
// Session. The child runs as a fully independent Session — it has its own
// history, its own Turns, its own Permissions — but the link lets the TUI
// render it as a nested footer beneath the parent Turn that opened it.
//
// Lifecycle parity with Permission: a SubAgentLink is created at spawn
// time and resolved when the child Session finishes. Once resolved, the
// link is immutable.
type SubAgentLink struct {
	ID            SubAgentID
	ParentSession SessionID
	ParentTurn    TurnID
	ChildSession  SessionID
	Prompt        string
	State         SubAgentState
	Verdict       Verdict // empty until State=Resolved
	SpawnedAt     time.Time
	ResolvedAt    time.Time // zero until State=Resolved
}

// SubAgentState is closed: a link is either live (child still running) or
// resolved (child terminated, verdict attached).
type SubAgentState string

const (
	SubAgentLive     SubAgentState = "live"
	SubAgentResolved SubAgentState = "resolved"
)

// withResolution returns a copy of the link with the given verdict and
// resolution timestamp. Receiver untouched.
func (s SubAgentLink) withResolution(v Verdict, now time.Time) SubAgentLink {
	s.State = SubAgentResolved
	s.Verdict = v
	s.ResolvedAt = now
	return s
}

File: tui/src/core/domain/subagent.ts
// L1: SubAgentLink with typestate.
//
// Mirrors the agentcore SubAgentLink: a live link is one whose child
// Session has not yet finished. Once resolved, the verdict is attached
// immutably.

import { type SessionID, type SubAgentID, type TurnID } from "./ids.js";
import { type Verdict } from "./verdict.js";

export type SubAgentState = "live" | "resolved";

type SubAgentLinkBase<S extends SubAgentState> = {
  readonly id: SubAgentID;
  readonly parentSession: SessionID;
  readonly parentTurn: TurnID;
  readonly childSession: SessionID;
  readonly prompt: string;
  readonly spawnedAt: Date;
  readonly state: S;
};

export type SubAgentLive = SubAgentLinkBase<"live">;

export type SubAgentResolved = SubAgentLinkBase<"resolved"> & {
  readonly verdict: Verdict;
  readonly resolvedAt: Date;
};

export type SubAgentLink<S extends SubAgentState = SubAgentState> =
  S extends "live" ? SubAgentLive :
  S extends "resolved" ? SubAgentResolved :
  SubAgentLive | SubAgentResolved;

export function isLive(l: SubAgentLink): l is SubAgentLive {
  return l.state === "live";
}

export function isResolvedSubAgent(l: SubAgentLink): l is SubAgentResolved {
  return l.state === "resolved";
}
